### ISSUE TITLE
Enhancement: [no-unused-vars] Add option `typeArgs` to allow changing how unused generic are used

### ISSUE THREAD

**Issue by RebeccaStevens** [typescript-eslint/typescript-eslint#6720]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unused-vars/

### Description

The rule currently only has the options inherited from the base rule.
This new option is specifically for the TypeScript feature of type arguments/generics.

For the following use case, setting this new option to "none" would stop any issues being reported.

### Fail

```typescript
// With current `args: "none"` option, type arguments are unaffected.

declare global {
  interface ReadonlyArray<T> {
                       // ^ `T` is technically unused, but it must still be specified.
    at<U extends number>(index: U): this[U];
    at<R>(this: readonly [...any[], R], index: -1): R;
  }

  interface Array<T> {
               // ^ Again `T` is technically unused, but it must still be specified.
    at<U extends number>(index: U): this[U];
    at<R>(this: readonly [...any[], R], index: -1): R;
  }
}
```

### Pass

```typescript
// With new `typeArgs: "none"` option, type arguments no longer report issues.

declare global {
  interface ReadonlyArray<T> {
    at<U extends number>(index: U): this[U];
    at<R>(this: readonly [...any[], R], index: -1): R;
  }

  interface Array<T> {
    at<U extends number>(index: U): this[U];
    at<R>(this: readonly [...any[], R], index: -1): R;
  }
}
```

**Comment by bradzacher** [comment]
Thus has been brought up before in the context of your specific code sample because those are technically necessary unused variables - but the reality is that the times that you actually want to have an unused type parameter are the edge case, not the common case.

For other tyoes of things there are dubious, but valid times you might like an unused thing:
- `args` has a `none` because it is not uncommon to have unused arguments to denote the signature of an overridden or interface method.
- `caughtErrors` has a `none` because not all codebases can syntactically support variable-less catch.

But in reality for type parameters the only time you physically need an unused one is when declaration merging.

The only time you might want one is if you're doing some advanced and really rare stuff with storing types to be extracted via conditional types later.

Even if you have those cases in your codebase - they're still going to be the exception - not the rule - so you're not going to want to turn off all type parameter checking for them.

Instead our advice is to use a disable comment to suppress this intentionally unused case and also to clearly document to others why its unnecessary.
